src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs

```diff
@@ -1125,4 +1127,132 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
             codeStream.Emit(ILOpcode.call, emitter.NewToken(helper));
         }
     }
+
+    class CustomTypeMarshaller : Marshaller
+    {
+        protected override void TransformManagedToNative(ILCodeStream codeStream)
+        {
+            var marshallerType = MarshalAsDescriptor.CustomMarshallerType;
+            if (marshallerType == null)
+            {
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            if (marshallerType.IsGenericDefinition)
+            {
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            var customMarshallerType = Context.SystemModule.GetKnownType("System.Runtime.InteropServices", "ICustomMarshaler");
+
+            if (!marshallerType.CanCastTo(customMarshallerType))
+            {
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            if (ManagedParameterType.IsPointer || ManagedParameterType.IsValueType)
+            {
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
```

Comment:
Would be nice to deduplicate as much of this as possible with TransformNativeToManaged.